index.rst

```diff
@@ -1,60 +1,850 @@
+
+- to fix a problem, or just get something running at all for the first time;
+- to run the same pipeline on more data IDs;
+- to run additional tasks;
+- any combination of the above.
+
+There are no general rules about what happens when the user revisits one of the previous steps after performing a later one; each case is different and needs to be thought through carefully.
+In some cases, we may need to rely on the user for extra information: for example, if the user changes a configuration option after generating the `~lsst.pipe.base.QuantumGraph`, do we need to regenerate it?
+Or can we just re-run the existing graph?
+At present, there's no way for the software to tell whether a configuration (or software change, for that matter) would affect the graph; we *must* rely on the user.
+
+There is also at least one case where users have good reasons to prefer different orders of operations, even if starting from the beginning:
+
+- Users who just want to get something working will generally want to build a `~lsst.pipe.base.QuantumGraph` before creating an output collection and writing/checking provenance, to fail as early as possible (and avoiding writing anything to the repo).
+
+- Users who expect to run multiple `QuantumGraphs<lsst.pipe.base.QuantumGraph>` in a campaign while writing results to the same output collection (especially in batch contexts) will often want to create that collection up front to avoid race condition.
+
+Finally, users in at least some contexts have a strong expectation that they will be able to perform *all* of these steps (or, rather, arbitrary subsets!) with a single command-line invocation.
+This design mostly attempts to meet that expection, by mapping steps to keyword arguments/command-line options *as well as* methods/subcommands.
+For example, one can use `Campaign.edit` to set the input collections (`collections.inputs`) without updating anything downstream, but also use the same :option:`--input` option in `Campaign.run` to change them at that stage (or set them for the first time if starting from scratch).
+
+It's worth questioning whether the right design decision is to instead [try to] push back on that single-invocation expectation in the name of simplicity; that's just not something I've done here (and I suspect at least a couple of single-invocation use cases are really quite well-motivated).
+
+.. [#comparing-campaigns]  The use case of running similar-but-not-identical pipelines on the same data IDs in order to compare their outputs is intentionally *not* included here, because that isn't something that should be done within one campaign; this is a use case best handled by using a different campaign for each pipeline (and possibly `importing<Campaign.import_quantum_graph>` a `~lsst.pipe.base.QuantumGraph`).
+
+
+The Collection Stack
+--------------------
+
+The collections associated with a campaign are organized largely as a stack - in the first-in, first-out data structure sense.
```

Comment:
Um, a stack is LIFO.

Reply:
Hahaha, yup.  I guess I'm a little less embarrassed that it's just this sentence that's wrong rather than the rest of the sentence.